### PR TITLE
Theo/dosed4

### PR DESCRIPTION
**Pull request by TheoMoutakanni**
- Add the possibility to perform a spectrogram on the signal
- New model DOSED4 to use both raw signals and spectrograms

### CODE REVIEW

File: dosed/datasets/dataset.py [hunk 1]
```diff
@@ -410,22 +407,62 @@ def __init__(self,
         )
         self.ratio_positive = ratio_positive
 
+        self.index_to_record_event = []
```
Comment (vthorey) [inline dosed/datasets/dataset.py]: C'est pas bon ça j'ai l'impression que c'est l'ancienne version du get sample n on ?

File: tests/test_dataset.py [hunk 2]
```diff
@@ -41,7 +41,14 @@ def signals():
                         "min_value": -150,
                     "max_value": 150,
                 }
-            }
+            },
+            'spectrogram': {
+                "nperseg": 8,
+                "nfft": 8,
+                "downsampling_t": 1,
```
Comment (vthorey) [inline tests/test_dataset.py]: _t _f ? -> replace by temporal frequential

File: dosed/datasets/dataset.py [hunk 3]
```diff
@@ -105,11 +108,33 @@ def __init__(self,
         data = Parallel(n_jobs=n_jobs, prefer="threads")(delayed(get_data)(
             filename="{}/{}".format(h5_directory, record),
             signals=signals,
-            fs=fs
-        ) for record in self.records)
+            fs=fs,
+            window=self.window
+        ) for record in tqdm.tqdm(self.records))
```
Comment (vthorey) [inline dosed/datasets/dataset.py]: tqdm in parallel ? Does it output correct outputs in parallele mode ?

File: dosed/datasets/dataset.py [hunk 4]
```diff
@@ -105,11 +108,33 @@ def __init__(self,
...
+
+        ##################
+        # Set all variables of the dataset
+        set_fs = set([d.pop('fs', None) for d in data])
+        assert len(set_fs) == 1, set_fs
+        self.fs = set_fs.pop()
+
+        set_window_size = set([d.pop('window_size', None) for d in data])
```
[the remark this answers is not in the document]
Comment (TheoMoutakanni) [reply dosed/datasets/dataset.py]: I reworked that part to be clearer

File: dosed/datasets/dataset.py [hunk 5]
```diff
@@ -105,11 +108,33 @@ def __init__(self,
...
+
+        if self.number_of_channels["spec"] > 0:
+            self.input_shape["spec"] = (self.number_of_channels["spec"],
+                                        data[0]["spec"].shape[-2],
+                                        self.window_size)
+        ##################
 
         for record, data in zip(self.records, data):
-            signal_size = data.shape[-1]
+            signal_size = data.pop('signal_size', None)
```
[the remark this answers is not in the document]
Comment (TheoMoutakanni) [reply dosed/datasets/dataset.py]: Done

File: dosed/datasets/dataset.py (hunk 1 above)
Comment (TheoMoutakanni) [reply dosed/datasets/dataset.py]: I changed the place of this function but it's the good version of it

File: dosed/datasets/dataset.py [hunk 6]
```diff
@@ -89,10 +89,11 @@ def __init__(self,
             get_data = memory.cache(get_h5_data)
             get_events = memory.cache(get_h5_events)
 
-        self.window_size = int(self.window * self.fs)
-        self.number_of_channels = len(signals)
+        self.number_of_channels = {
+            "spec": len([signal for signal in signals if "spectrogram" in signal]),
```
[the remark this answers is not in the document]
Comment (TheoMoutakanni) [reply dosed/datasets/dataset.py]: Done

File: tests/test_dataset.py (hunk 2 above)
Comment (TheoMoutakanni) [reply tests/test_dataset.py]: Done

File: dosed/datasets/dataset.py (hunk 3 above)
Comment (TheoMoutakanni) [reply dosed/datasets/dataset.py]: No it doesn't.
It only shows the time when a record starts being processed and not the time when it is finished.
I will find a workaround.

**Comment by vthorey** [comment]
These changes might be itegrated later on

**Pull request closed without merging**